### ISSUE TITLE
Recursive mode

### ISSUE THREAD

**Issue by sevagh** [wata727/tflint#132]
Hello,

I'm thinking of working on and submitting a pull request for a feature to recurse into directories, e.g. `tflint -r .` to visit all of my subdirectories of Terraform recipes and lint each one.

Is that a good feature? Any pros/cons to implementing such a thing?

**Comment by wata727** [comment]
What kind of situation is assumed for this feature?

In my opinion, since Terraform covers only the current directory, I think that further features are not unnecessary.

**Comment by sevagh** [comment]
If you do `terraform fmt` or `terraform fmt --diff` from a top-level directory, it will descend into subdirectories and format.

I want to lint an entire collection of Terraform recipes:

```

**Comment by wata727** [comment]
I see, I agree with your opinion. But, unlike `terraform fmt`, TFLint has various settings according to the analysis target. I think that this property is similar to `terraform plan`.

My recommendation is to run TFLint each time after `terraform plan` and before `terraform apply`.

**Comment by sevagh** [comment]
OK, fair enough.

**Comment by jclarksnps** [comment]
I actually thing this request was completely valid.  Imagine running tflint on a git repo before commit... why would i want to specify each module and not run it recursively from the root.  @wata727

**Comment by charlex** [comment]
I'm also a fan of this idea.

Because I don't have my tf files in the root of my repo, my CI code must have an extra `terraform init`. 

This is what I must do in CI to accommodate this constraint

```bash
terraform init -input=false tf
terraform fmt -write=false -diff tf
cd tf
terraform init -input=false
tflint
```

Instead of the simpler:

```bash
terraform init -input=false tf
terraform fmt -write=false -diff tf
tflint tf
```

It seems like the closer `tflint` gets to matching the syntax of `terraform`, the better things will be for everyone! Perhaps it might even make tflint a candidate for inclusion in the core `terraform`. 

Do the maintainers of this project have plans to go in this direction? Or open to this idea?

**Comment by wata727** [comment]
@charlex v0.8.2 now accepts a directory as an argument. This feature will help to avoid the redundant `cd`. In this case, please note that you may need to specify the directory generated by `terraform init` setting [`TF_DATA_DIR`](https://www.terraform.io/docs/commands/environment-variables.html#tf_data_dir). See also #296 

If you face any troubles, please open a new issue because this issue relates to the recursive inspection. Thanks.

@jclarksnps Sorry for the late response. And I think about this feature again. However, it is still difficult to implement this feature. Even if the recursive inspection is possible, I think that it is not convenient because `terraform init` is required in each directory.

However, I also know from feedbacks that we need such features. If anyone has a good idea, I may be working on it in the future.

**Comment by charlex** [comment]
That's great @wata727! Thank you for noting both the new feature, and the `TF_DATA_DIR` which I wasn't aware of and helps clean up some of my CI config.

I am new to TF in general and was on TFLint v0.8.1, and am excited to see the addition of directory support in v0.8.2!

**Comment by trallnag** [comment]
Hey @wata727,

> I think that it is not convenient because terraform init is required in each directory.

Why is `terraform init` necessary? From what I understand following should be enough:

```
# In root module
terraform init
# In the root module, nested modules, and cached modules
tflint .
tflint modules/abc
tflint .teraform/modules/xyz
```

**Comment by bendrucker** [comment]
If child modules source modules from anything other than a relative local path, you must run `init` there before TFLint can run in that working directory.

As this issue has been closed for a long time and there are no plans to introduce this feature, I'm closing it to further replies.
